### PR TITLE
Fix sampling for version multiplying factor

### LINKED ISSUE #21287: QRCodeDetector::decode error: (-4:Insufficient memory) Failed to allocate ... bytes (tens of GBs)
- OpenCV => 4.5.4 (not specifically)

`QRCodeDetector::decode` (also `QRCodeDetector::detectAndDecode`) crashes trying to allocate an unreasonable amount of memory.

`cv2.error: OpenCV(4.5.4) D:\a\opencv-python\opencv-python\opencv\modules\core\src\alloc.cpp:73: error: (-4:Insufficient memory) Failed to allocate 54056250000 bytes in function 'cv::OutOfMemoryError'`

##### Steps to reproduce

Use the attached QR code
```python
import cv2 as cv
im = cv.imread("hIfAR.png")
det = cv.QRCodeDetector()
(rv, points) = det.detect(im) # succeeds
(data, rectified) = det.decode(im, points) # fails
# (data, points, rectified) = det.detectAndDecode(im) # also fails
```

The issue was brought to my attention by [pepoluan on Stack Overflow](https://stackoverflow.com/questions/70373514/opencv-qrcodedetector-out-of-memory-even-with-a-small-file), who allows the code to be a test case. The code contains a zip file of a json file.

> the information within the Zip file is dummy info, so it's okay to be public.

hIfAR

### PR DESCRIPTION
**Pull request by AleksandrPanov**
fixes #21287

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -2367,22 +2436,49 @@ bool QRDecode::versionDefinition()
...
 
     const int delta_rows = cvRound((postIntermediate.rows * 1.0) / version_size);
     const int delta_cols = cvRound((postIntermediate.cols * 1.0) / version_size);
 
     vector<double> listFrequencyElem;
-    for (int r = 0; r < postIntermediate.rows; r += delta_rows)
+    for (int i = 0, r = 0; i < version_size; i++ && r < postIntermediate.rows, r+= delta_rows)
     {
-        for (int c = 0; c < postIntermediate.cols; c += delta_cols)
+        for (int j = 0, c = 0; j < version_size; j++ && c < postIntermediate.cols, c+= delta_cols)
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:2481]: Overcomplicated (need to rewrite) or just invalid "for" loops.

File: modules/objdetect/src/qrcode.cpp [hunk 2]
```diff
@@ -2369,7 +2369,8 @@ bool QRDecode::samplingForVersion()
     CV_TRACE_FUNCTION();
     const double multiplyingFactor = (version < 3)  ? 1 :
                                      (version == 3) ? 1.5 :
-                                     version * (version + 1);
+                                     (version < 6) ? version*(version+1) :
+                                     version;
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:2373]: > `version;`

`version * 5` ?

File: modules/objdetect/src/qrcode.cpp [hunk 3]
```diff
@@ -2378,45 +2379,33 @@ bool QRDecode::samplingForVersion()
...
-    for (int r = 0; r < postIntermediate.rows; r += delta_rows)
-    {
-        for (int c = 0; c < postIntermediate.cols; c += delta_cols)
-        {
+    vector<int> deltas_rows(version_size, delta_rows);
+    vector<int> deltas_cols(version_size, delta_cols);
+
+    for (int i = 0; i < abs(skipped_rows); i++) {
+        deltas_rows[(i*version_size)/abs(skipped_rows)+(version_size/abs(skipped_cols)-1)/2]
+        += skipped_rows > 0 ? 1 : -1;
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:2390]: Please separate index calculation to simplify code.

File: modules/objdetect/src/qrcode.cpp [hunk 4]
```diff
@@ -2378,45 +2379,33 @@ bool QRDecode::samplingForVersion()
...
+    }
+    for (int i = 0; i < abs(skipped_cols); i++) {
+        deltas_cols[(i*version_size)/abs(skipped_cols)+(version_size/abs(skipped_cols)-1)/2]
+        += skipped_cols > 0 ? 1 : -1;
+    }
+
+    const double totalFrequencyElem = countNonZero(postIntermediate) / static_cast<double>(postIntermediate.total());
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:2397]: Looks like `postIntermediate` is not a binary (0/255) image. So there is no sense to use `countNonZero()` here.

Perhaps we want to use `norm(postIntermediate, NORM_L1 | NORM_RELATIVE);` here.

---

Do we have 0.5 value here?

Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp:2397]: actually `postIntermediate` is binary (0/255) image in this case:

1. `no_border_intermediate` is binary
2. `postIntermediate` = `resize(no_border_intermediate, newFactorSize, INTER_AREA)`
3. `newFactorSize` is integer and  InterpolationFlags is `INTER_AREA` so `postIntermediate` is binary (0/255) image

File: modules/objdetect/src/qrcode.cpp (hunk 2 above)
Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp:2373]: fixing the tail processing process allows to reduce `multiplyingFactor` to 1 and skip strange`resize()` step

I want to add a few more tests and skip this step or reduce the `multiplyingFactor` to 2, for example.

File: modules/objdetect/src/qrcode.cpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp]: fixed

File: modules/objdetect/perf/perf_qrcode_pipeline.cpp [hunk 5]
```diff
@@ -106,10 +106,11 @@ PERF_TEST_P_(Perf_Objdetect_QRCode_Multi, decodeMulti)
 INSTANTIATE_TEST_CASE_P(/*nothing*/, Perf_Objdetect_QRCode,
     ::testing::Values(
         "version_1_down.jpg", "version_1_left.jpg", "version_1_right.jpg", "version_1_up.jpg", "version_1_top.jpg",
-        "version_5_down.jpg", "version_5_left.jpg", "version_5_right.jpg", "version_5_up.jpg", "version_5_top.jpg",
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/objdetect/perf/perf_qrcode_pipeline.cpp:109]: fixed

File: modules/objdetect/src/qrcode.cpp (hunk 2 above)
Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp]: now `multiplyingFactor`:
```
    const double multiplyingFactor = (version < 3)  ? 1. :
                                     (version == 3) ? 2. :
                                     3.;
```

File: modules/objdetect/src/qrcode.cpp (hunk 3 above)
Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp]: fixed

Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp]:
```
    for (int i = 0; i < abs(skipped_rows); i++) {
        // fix deltas_rows at each skip_step
        const double skip_step = static_cast<double>(version_size)/abs(skipped_rows);
        const int corrected_index = static_cast<int>(i*skip_step + skip_step/2);
        deltas_rows[corrected_index] += skipped_rows > 0 ? 1 : -1;
    }
```

**Comment by AleksandrPanov** [comment]
@alalek, fixed the issues

**Review by alalek — APPROVED** [review]

**Pull request merged**